File: src/main/java/de/hanno/hpengine/engine/graphics/query/GLQuery.java
package de.hanno.hpengine.engine.graphics.query;

import de.hanno.hpengine.renderer.GraphicsContext;

import static org.lwjgl.opengl.GL11.GL_TRUE;
import static org.lwjgl.opengl.GL15.GL_QUERY_RESULT_AVAILABLE;
import static org.lwjgl.opengl.GL33.glGetQueryObjectui64;

public interface GLQuery<RESULT> {
    void begin();

    void end();

    default boolean resultsAvailable() {
        return GraphicsContext.getInstance().calculate( () -> glGetQueryObjectui64(getQueryToWaitFor(), GL_QUERY_RESULT_AVAILABLE)) == GL_TRUE;
    }

    int getQueryToWaitFor();

    RESULT getResult();
}

File: engine/src/main/kotlin/de/hanno/hpengine/graphics/query/GLQuery.kt
package de.hanno.hpengine.graphics.query

import de.hanno.hpengine.graphics.GpuContext
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL33

interface GLQuery<RESULT> {
    fun begin(): GLTimerQuery?
    fun end()
    fun resultsAvailable(gpuContext: GpuContext): Boolean = gpuContext.onGpu {
        GL33.glGetQueryObjectui64(
            queryToWaitFor,
            GL15.GL_QUERY_RESULT_AVAILABLE
        )
    }.toInt() == GL11.GL_TRUE

    val queryToWaitFor: Int
    val result: RESULT
}